README.rst
=================
LightLdapd README
=================

LightLdapd is a small and easy to manage LDAP server for small
home/classroom/school sized networks that can include thin clients. It
is designed to provide network wide users/groups/etc support to
supplement the excellent network dns/dhcp/tftp support provided by
dnsmasq.

It uses local PAM and NSS for authentication to export the local
passwd/group/shadow/etc system databases. This means it requires no special
separate ldap database to manage, and users/groups/etc can be managed normally
on the LightLdapd server machine. No special database setup or ldap management
tools are required, it just exports all the local users/groups/etc. Client
machines can then use libpam_ldapd and libnss_ldapd to have all the same
users/groups/etc as are on the machine where LightLdapd is running.

It can also run inside a minimal chroot environment, giving more isolation
from the host system, and explorting a completely different set of
users/groups possibly using a completely different PAM/NSS setup from inside
the chroot.

The code is small, clean and efficient C, leveraging of existing libraries as
much as possible. It uses libev for an efficient event loop. It uses asn1c to
auto-generate the LDAP message parsing/generating from the ASN.1 spec. It uses
libpam or crypt for authentication. It uses mbedtls for TLS support. It is
small and efficient enough to run on a router or NAS.

LightLdapd was forked from the excellent entente by Sergey Urbanovich
with his blessing. The choice to fork was made in order to leave
entente as simple as possible and avoid adding features needed for
LightLdapd. Improvements have and will be fed back into entente when
they don't increase its size.

Status
======

LightLdapd is now functional enough to be used for clients using
pam_ldap authentication and/or nss_ldap access to passwd, group, and
shadow data. It now has TLS support so it should be safe to use on
untrusted networks.

The TLS support is relatively new but complete. If TLS is enabled it will
require it be started for bind operations. This should prevent clients from
accidentally using it insecurely on an untrusted network.

It does not yet have any logging, which can make it hard to figure out
what is wrong when something is not working.

See the github issue tracker for the most up to date status on what
is/isn't working.

Contents
========

.. This should be a brief description of the contents of the
   distribution. It should include a list of important features in a
   table like this;

==================== ======================================================
Name                 Description
==================== ======================================================
`<README.rst>`_      This file.
`<DESIGN.rst>`_      Details of the design.
`<DEVELOPMENT.rst>`_ Instructions for developers.
`<LICENSE>`_         Copyright and Licencing details.
`<NEWS.rst>`_        Summary of fixes and changes for each release.
`<TODO.rst>`_        List of outstanding tasks and future plans.
`<ldap.asn1>`_       The asn1 ldap protocol specification.
`*.[ch]`             The project source code.
`<chroot/>`_         Minimal chroot for testing.
==================== ======================================================

.. It wouldn't hurt to have a few paragraphs here suggesting were to
   look in the distribution for bits and pieces.


Install
=======

Dependencies
------------

* `asn1c <https://github.com/vlm/asn1c>`_ - `apt-get install asn1c`
* `libev <http://software.schmorp.de/pkg/libev.html>`_ - `apt-get install libev-dev`
* `libpam <http://www.kernel.org/pub/linux/libs/pam/>`_ - `apt-get install libpam-dev`
* `mbedtls <https://tls.mbed.org/>`_ - `apt-get install libmbedtls-dev`

Build
-----

To compile and install::

    make
    make install

Or (for building debian package)::

    make debian

Usage
=====

.. Simple Instructions for usage after installing. May include a
   reference to man pages or documentation in doc/, or USAGE

::

    lightldapd [options]

Or::

    /etc/init.d/lightldapd start
    # config file: /etc/default/lightldapd

Options
-------

-a  Allow anonymous access.
-b basedn  Set the basedn for the ldap server (default: "dc=lightldapd").
-r rootuser  Bind user for 'root' access to shadow data (default: "root").
-l  Bind to the loopback interface only.
-p port  Set local port number (default: 389).
-d  Run as a daemon.
-u runuser  Optional user to run as after dropping root privileges.
-R chroot  Optional path to chroot() into.
-C crtpath  Optional path to an ssl cert to use for TLS.
-A ca-path  Optional path to a ca-chain to use for TLS.
-K keypath  Optional path to a private key to use for TLS.
-U uids  Optional comma-separated list of uids or uid-ranges to export
  (default: "1000-29999").
-G gids  Optional comma-separated list of gids or gid-ranges to export
  (default: "100,1000-29999").
-N  Use NSS and crypt for authenticating instead of PAM.

Note lightldapd must run as root to open the default ldap serving
port, but using ``-u runuser`` it will use setuid() to drop root
privileges after starting. However, this also usually means it cannot
access nss shadow data so will not serve shadowAccount data. Clients
using pam_ldap for authentication don't need to access shadow data
anyway, and it is more secure to not export it.

However, if you want clients to use normal pam_unix authentication and
read shadow data using nss_ldap, then the runuser needs to be
privilged enough to read nss shadow data on the server. If the server
is using nss_unix, this is often done by adding the runuser to a
``shadow`` group that has read access to ``/etc/shadow``. You need to
also configure nss_ldap on the client machines to bind as the rootuser
with the ``rootbinddn`` setting so root (and only root)on the clients
can read shadow data.

Using ``-R chroot`` means lightldapd can be run in a chroot isolated from the
host system with a completely different NSS/PAM setup and users. The chroot
must include everything to configure NSS and PAM correctly, including all the
required PAM modules and libraries. The runuser, and cert paths are resolved
and read before switching to the chroot, so must exist on the host system.
Logging with syslog is initialized before switching to the chroot so it will
log to the host system and doesn't need anything configured in the chroot. The
runuser should also exist in the chroot with the same uid, and must have
permissions configured correctly for access to shadow data. The rootuser is
resolved after switching to the chroot, so only needs to exist in the chroot
environment.

Using ``-N`` means authentication is done using NSS and crypt against the the
shadow passwords instead of using PAM. This requires that the runuser have
read access to NSS shadow data. This is particularly useful for running in a
chroot, as it means the chroot doesn't need any pam modules installed, and
only requires a minimal nss setup with resolv.conf and passwd/group/shadow
files.

To enable TLS support you specify a cert file with the ``-C`` option, and
optionally a certificate authority chain file with the ``-A`` argument and/or
a separate private key file with the ``-K`` argument. If you don't use the
``-K`` option, the cert file must be a ``*.pem`` file containing both the cert
and private key. The file contining the private key must be readable by the
user lightldapd is started as, but doesn't have to be readable by the ``-u
runuser`` user. Typically it is set readable only by root. It is important to
configure your clients to use TLS and trust the cert used. If you are using
self-signed certs this typically means giving them a copy of the public cert.

To only expose a subset of your local uids or gids over ldap, use the `-U` and
`-G` options, setting them to a comma-separated list of ids or id-ranges to
include. The defaults are `-U 1000-29999` and `-G 100,1000-29999`. This
ensures that system users and groups are not exported, as clients typically
define their own system users and groups in their own /etc/passwd and
/etc/group configs. Note that authentication for bind is done against the host
(or chroot) system's users and is not limited to the `-U` exported uid ranges.
This means it is possible to bind as a user not exported by the LDAP server.
This means you can make the rootuser a special system user (uid < 1000) while
only exporting normal users (uid >= 1000).

Example usage with lighttpd
---------------------------

lighttpd.conf::

    server.modules += ( "mod_rewrite" )

    auth.backend = "ldap"
    auth.backend.ldap.hostname = "localhost"
    auth.backend.ldap.filter   = "(user=$)"

    auth.require = (
        "/tratata" => (
            "method"  => "basic",
            "realm"   => "lightldapd",
            "require" => "user=kiki|user=ooki"
        ),
    )



Support
=======

.. This should list all the user-level contact points for support,
   including mailing lists, discussion forums, online documentation,
   trackers, etc. It should also include instructions or pointers to
   instructions on procedures and conventions when using them.

Documentation
-------------

http://github.com/dbaarda/LightLdapd
  The project homepage.

http://minkirri.apana.org.au/wiki/LightLdapd
  An early brainstorming wiki before the github project was created.

Discussion
----------

.. Provide links to any IRC channels, mailing lists or online
   discussion forums, giving any necissary subscription information
   etc.

Reporting Problems
------------------

.. This should describe the procedure for users to report bugs,
   providing any useful links.

File any problems/bugs/suggestions/questions on the github issue
tracker.

Development
===========

See `<DEVELOPMENT.rst>`_ for development instructions including debugging.

See `<DESIGN.rst>`_ for general design philosophy and ideas.

----

http://github.com/dbaarda/LightLdapd
$Id: README,v 65b64de6b1e1 2014/01/20 02:32:20 abo $


NEWS.rst
===============
LightLdapd NEWS
===============

This is a summary of user visible changes for each release. Where possible it
includes references to bug and patch numbers. Please ensure that things are
moved to here from the TODO file as they get done. Before doing a release
please ensure that this document is up to date.

Changes in 1.0.1 (Not yet released)
===================================

* Added `-R chroot` support.

  Add support for running in a chroot with -R chroot. Add initializing of
  syslog and basic start/stop logging to ensure logging is initialized before
  switching into the chroot. See README.rst for details.

* Made all `*.h` docstrings doxygen compatible.

* Moved buffer class into its own file and add tests.

  Move buffer_t class and methods into buffer.h from ldap_server.[ch] and add
  buffer_test.c.

Changes in 1.0.0 (released 2020-01-02)
======================================

* Forked LightLdapd project from entente.

  With permission and thanks to Sergey Urbanovich, the author of entente.

* Changed name from entente to lightldapd.

  The forked project is now named LightLdapd, and the binary and cfgs have
  been renamed to lightldapd.

* Changed license from MIT to GPLv3.

  This means we require contributions to come back rather than spawn private
  forks. I have confirmed with the entente author this is OK.

* Improve project documentation:

  Add documentation based on templates in
  http://minkirri.apana.org.au/~abo/projects/prjdocs/.

* Tidy code.

  Reformat again using a different preferred style without tabs. Change `make
  tidy` target to reformat using tidyc tool.

* #9,#10 Improve design.

  Restructured using ldap_server, ldap_connection, ldap_request, ldap_reply
  structs, copying the design of https://github.com/taf2/libebb.

* Extend Search support.

  Extended search support enough to support libnss-ldap clients, exporting the
  local nsswitch view of passwd/group/etc.

  #3 Add support for typesOnly and attribute selection.

* #2 Optimize Search.

  Added Filter_scope() analysis to figure out what the search is constrained
  to instead of scanning everything.

* #4 Add StartTLS support.

  StartTLS support with security checking before allowing bind implemented
  using mbedtls.

* #13 Make served users/groups configurable.

  Support serving only some user/group ranges using `-U` and `-G` arguments to
  specify uid/gid ranges to export.


Changes in entente 1.1 (merged 2014-01-25)
==========================================

* Improve options.

  Make -b basedn argument work. Make -l loopback argument work. Simplified and
  removed environment based settings.

* Make pam authentication failures non-blocking:

  It will nolonger stall all connections ~2 seconds whenever someone tries a
  bad user/passwd. Instead only the connection that failed to bind is paused
  for time configured by pam_fail_delay.

* Improved bind result failure returncodes.

  Changed the bind failure response resultcode from "other" to
  "invalidDNSyntax" or "invalidCredentials", depending on why it failed.

* Tidy code.

  Reformated using "indent -linux -l120". Added a "tidy" make target to do
  this automatically.

  Tidied up lots of code, simplifying methods, renaming things to be more
  consistant, and make better use of common library functions.

  Made all memory allocation checking to use "alloc or die" macro's, since
  libev will abort on alloc failures anyway.

* Improved/updated debian build:

  Updated for debhelper v9. Added debclean make target.


Changes in entente 1.0 (committed 2011-04-11)
=============================================

* Initial Release

  Published at https://github.com/urbanserj/entente. Does enough to work as a
  pam_ldap auth server.


----

$Id: TODO,v 1.40 2004/10/18 02:30:53 abo Exp $


main.c
/*=
 * Copyright (c) 2017 Donovan Baarda <abo@minkirri.apana.org.au>
 * Based on entente Copyright (c) 2010, 2011 Sergey Urbanovich
 *
 * Licensed under the GPLv3 License. See LICENSE file for details.
 */

#include "utils.h"
#include "ldap_server.h"
#include "ranges.h"
#include "pam.h"
#include "log.h"
#include <unistd.h>
#include <syslog.h>

char *setting_port = "389";
bool setting_loopback = 0;
bool setting_authnss = 0;
bool setting_daemon = 0;
char *setting_runuser = "root";
char *setting_chroot = NULL;
char *setting_basedn = "dc=lightldapd";
char *setting_rootuser = "root";
bool setting_anonok = 0;
char *setting_crtpath = NULL;
char *setting_caspath = NULL;
char *setting_keypath = NULL;
char *setting_uids = "1000-29999";
char *setting_gids = "100,1000-29999";
void settings(int argc, char **argv);

int main(int argc, char **argv)
{
    ev_loop *loop = EV_DEFAULT;
    mbedtls_net_context socket;
    ldap_server server;
    char *server_addr;
    uid_t runuid;
    ldap_ranges uids, gids;

    settings(argc, argv);
    log_init("lightldapd", setting_daemon, LOG_DEBUG);
    lnote("lightldapd starting");
    server_addr = setting_loopback ? "127.0.0.1" : NULL;
    runuid = name2uid(setting_runuser);
    if (!ldap_ranges_init(&uids, setting_uids))
        lerrx(EX_USAGE, "Invalid -U value: \"%s\"", setting_uids);
    if (!ldap_ranges_init(&gids, setting_gids))
        lerrx(EX_USAGE, "Invalid -G value: \"%s\"", setting_gids);
    if (ldap_server_init
        (&server, loop, setting_basedn, setting_rootuser, setting_anonok, setting_crtpath, setting_caspath,
         setting_keypath, &uids, &gids))
        lerr(1, "ldap_server_init() failed");
    if (mbedtls_net_bind(&socket, server_addr, setting_port, MBEDTLS_NET_PROTO_TCP))
        lerr(1, "mbdedtls_net_bind() failed");
    if (setting_daemon && daemon(1, 0))
        lerr(1, "daemon() failed");
    if (setting_chroot && chroot(setting_chroot))
        lerr(1, "chroot() failed");
    if (chdir("/"))
        lerr(1, "chdir() failed");
    if (runuid && setuid(runuid))
        lerr(1, "setuid() failed");
    if (setting_authnss)
        auth_user = auth_nss;
    ldap_server_start(&server, socket);
    ev_run(loop, 0);
    lnote("lightldapd stopping");
    return 0;
}

void settings(int argc, char **argv)
{
    int c;

    while ((c = getopt(argc, argv, "ab:dlp:r:u:A:C:G:K:NR:U:")) != -1) {
        switch (c) {
        case 'a':
            setting_anonok = true;
            break;
        case 'b':
            setting_basedn = optarg;
            break;
        case 'd':
            setting_daemon = true;
            break;
        case 'l':
            setting_loopback = true;
            break;
        case 'p':
            setting_port = optarg;
            break;
        case 'r':
            setting_rootuser = optarg;
            break;
        case 'u':
            setting_runuser = optarg;
            break;
        case 'A':
            setting_caspath = optarg;
            break;
        case 'C':
            setting_crtpath = optarg;
            break;
        case 'G':
            setting_gids = optarg;
            break;
        case 'K':
            setting_keypath = optarg;
            break;
        case 'N':
            setting_authnss = true;
            break;
        case 'R':
            setting_chroot = optarg;
            break;
        case 'U':
            setting_uids = optarg;
            break;
        default:
            fprintf(stderr,
                    "Usage: %s [-a] [-b dc=lightldapd] [-r rootuser] [-l] [-p 389] [-d] \\\n"
                    "  [-u runuser] [-R chroot] [-C crtfile] [-A ca-file] [-K keyfile] \\\n"
                    "  [-U 1000-29999,...] [-G 100,1000-29999,...] [-N]", argv[0]);
            exit(EX_USAGE);
        }
    }
}
